crates/holochain_p2p/src/spawn/actor.rs

```diff
@@ -879,9 +901,96 @@ impl HolochainP2pActor {
             kitsune,
             pending,
             request_duration_metric: create_p2p_request_duration_metric(),
+            pruning_task_abort_handle,
         }))
     }
 
+    // Prunes expired URLs at an interval and checks the peer store for agent infos of unresponsive
+    // URLs. If there is an updated agent info since the URL was marked unresponsive, the URL will
+    // be pruned.
+    fn spawn_pruning_task(
+        interval_ms: u64,
+        kitsune2: DynKitsune,
+        db_getter: GetDbPeerMeta,
+    ) -> AbortHandle {
+        tokio::spawn(async move {
+            loop {
+                tokio::time::sleep(Duration::from_millis(interval_ms)).await;
+
+                let spaces = kitsune2.list_spaces();
+                let pruning_futs =
+                    spaces.into_iter().map(|space_id| {
+                        let db_getter = db_getter.clone();
+                        let kitsune2 = kitsune2.clone();
+                        async move {
+                            let space = kitsune2.clone().space(space_id.clone()).await?;
+                            let peer_store = space.peer_store().clone();
+                            let db = db_getter(DnaHash::from_k2_space(&space_id)).await?;
+                            // Prune any expired entries.
+                            db.write_async(|txn| -> DatabaseResult<()> {
+                                let prune_count = txn.execute(sql_peer_meta_store::PRUNE, [])?;
+                                tracing::debug!("pruned {prune_count} rows from meta peer store");
+                                Ok(())
+                            })
+                            .await
+                            .map_err(HolochainP2pError::other)?;
+
+                            // Get agent infos from peer store and compare if there are any up to date ones with
+                            // any of the unresponsive URLs.
+                            // That would indicate that the URL was unresponsive temporarily and has become
+                            // responsive again.
+                            let agents = peer_store.get_all().await?;
```

Comment:
I could add a method `get_by_url` to the peer store. We're probably going to use it in several places.

Reply:
I've added the method, but it's not helpful for this use case. I'd need to return all unresponsive URLs with timestamps from the database query and then for each of the URLs `get_by_url` from the peer store and then compare the timestamps. Doesn't read well, will access the peer store multiple times, in comparison to having a few hundred max agent infos in memory for a moment.